src/Stratis.Bitcoin/P2P/Peer/NetworkPeerClient.cs

```diff
@@ -0,0 +1,172 @@
+        private AsyncLock writeLock;
+
+        /// <summary>Stream to send and receive messages through established TCP connection.</summary>
+        /// <remarks>Write operations on the stream have to be protected by <see cref="writeLock"/>.</remarks>
+        public NetworkStream Stream { get; private set; }
+
+        /// <summary>Task that cares about the client once its connection is accepted up until the communication is terminated.</summary>
+        /// <remarks>The client is being processed as long as this task is not complete.</remarks>
+        public TaskCompletionSource<bool> ProcessingCompletion { get; private set; }
+
+        /// <summary><c>1</c> if the instance of the object has been disposed or disposing is in progress, <c>0</c> otherwise.</summary>
+        private int disposed;
+
+        /// <summary>Address of the end point the client is connected to, or <c>null</c> if the client has not connected yet.</summary>
+        public IPEndPoint RemoteEndPoint
+        {
+            get
+            {
+                return (IPEndPoint)this.tcpClient?.Client?.RemoteEndPoint;
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new network client.
+        /// </summary>
+        /// <param name="Id">Unique identifier of a client.</param>
+        /// <param name="tcpClient">Initialized TCP client, which may or may not be already connected.</param>
+        /// <param name="loggerFactory">Factory for creating loggers.</param>
+        public NetworkPeerClient(int Id, TcpClient tcpClient, ILoggerFactory loggerFactory)
+        {
+            this.tcpClient = tcpClient;
+
+            this.loggerFactory = loggerFactory;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, string.Format("[{0}{1}] ", this.Id, this.RemoteEndPoint != null ? "-" + this.RemoteEndPoint.ToString() : ""));
+
+            this.Stream = this.tcpClient.Connected ? this.tcpClient.GetStream() : null;
+            this.ProcessingCompletion = new TaskCompletionSource<bool>();
+
+            this.writeLock = new AsyncLock();
+        }
+
+        /// <summary>
+        /// Connects the network client to the target server.
+        /// </summary>
+        /// <param name="endPoint">IP address and port to connect to.</param>
+        /// <param name="cancellation">Cancellation token that allows aborting the operation.</param>
+        /// <exception cref="OperationCanceledException">Thrown when the connection attempt was aborted.</exception>
+        public async Task ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation)
+        {
+            this.logger = this.loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.Id}-{endPoint}] ");
+            this.logger.LogTrace("({0}:'{1}')", nameof(endPoint), endPoint);
+
+            try
+            {
+                await Task.Run(async () => await this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port), cancellation);
```

Comment:
`this.tcpClient.ConnectAsync` does nto take a cancellation token that's why you create a task?
Will it even matter passing cancellation in such a way?
Can we not just rely on the tcp timeout?

Reply:
yes, this is sadly imperfect API that we can't do much about
this is recommended way of introducing cancellation token to connect here
it does matter a lot because when you shutdown, you don't want to wait 30 or whatever seconds
so yes, it does matter

Reply:
Ah ok if this is the recommended way to override cancellations then very good